homeassistant/components/azure_event_hub/__init__.py

```diff
@@ -1,89 +1,213 @@
+                vol.Optional(CONF_FILTER, default={}): FILTER_SCHEMA,
+            },
+            cv.has_at_least_one_key(
+                CONF_EVENT_HUB_CON_STRING, CONF_EVENT_HUB_NAMESPACE
+            ),
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-async def async_setup(hass: HomeAssistant, yaml_config: Dict[str, Any]):
+async def async_setup(hass, yaml_config):
     """Activate Azure EH component."""
     config = yaml_config[DOMAIN]
-
-    event_hub_address = (
-        f"amqps://{config[CONF_EVENT_HUB_NAMESPACE]}"
-        f".servicebus.windows.net/{config[CONF_EVENT_HUB_INSTANCE_NAME]}"
-    )
-    entities_filter = config[CONF_FILTER]
-
-    client = EventHubClientAsync(
-        event_hub_address,
-        debug=True,
-        username=config[CONF_EVENT_HUB_SAS_POLICY],
-        password=config[CONF_EVENT_HUB_SAS_KEY],
+    entities_filter = config.get(CONF_FILTER)
+    if config.get(CONF_EVENT_HUB_CON_STRING, None):
+        client_args = {"conn_str": config[CONF_EVENT_HUB_CON_STRING]}
+        conn_str_client = True
+    else:
+        client_args = {
+            "fully_qualified_namespace": f"{config[CONF_EVENT_HUB_NAMESPACE]}.servicebus.windows.net",
+            "credential": EventHubSharedKeyCredential(
+                policy=config[CONF_EVENT_HUB_SAS_POLICY],
+                key=config[CONF_EVENT_HUB_SAS_KEY],
+            ),
+            "eventhub_name": config[CONF_EVENT_HUB_INSTANCE_NAME],
+        }
+        conn_str_client = False
+
+    instance = hass.data[DOMAIN] = AEHThread(
+        hass, client_args, conn_str_client, entities_filter
     )
-    async_sender = client.add_async_sender()
-    await client.run_async()
-
-    encoder = JSONEncoder()
-
-    async def async_send_to_event_hub(event: Event):
+    instance.async_initialize()
+
+    async def async_unload_entry(hass: HomeAssistant):
+        """Shut down the AEH Thread."""
+        instance.queue.put(None)
+        instance.join()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_unload_entry)
+
+    instance.start()
+    return await instance.async_ready
+
+
+class AEHThread(threading.Thread):
+    """A threaded event handler class."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client_args: Dict[str, Any],
+        conn_str_client: bool,
+        entities_filter: vol.Schema,
+    ):
+        """Initialize the listener."""
+        threading.Thread.__init__(self, name="AzureEventHub")
+        self.hass = hass
+        self.queue = queue.Queue()
+        self._client_args = client_args
+        self._conn_str_client = conn_str_client
+        self._entities_filter = entities_filter
+        self.async_ready = asyncio.Future()
+        self.write_errors = 0
+        self.shutdown = False
+        # suppress the INFO and below logging on the underlying packages, they are very verbose, even at INFO
+        logging.getLogger("uamqp").setLevel(logging.WARNING)
```

Comment:
We don't want side effects in init method. Please move this either outside init or to a dedicated setup method.